Big Tetris Parties

bitoffdev opened the pull request:
TL;DR This patch allows more than the previous 2 players to compete, and
makes many minor improvements.

![Peek 2021-01-01 23-55](https://user-images.githubusercontent.com/5495776/103451285-0e121600-4c90-11eb-987b-33d0a05c5c45.gif)

---

- Add magic number to all messages to help catch any network issues
- Add message_type field to message wire header, and remove the message_type
  from the body of many messages
- Define new method, message_type_to_str, to print a human-readable message
  type for logging
- Create new helper method message_blob
- Declare a macro name "PLAYER_NAME_MAX_CHARS", and use it consistently to
  validate the player names.
- Refactor StringArray to take a max length and define the StringArray
  serialized header bytes using a struct
- Bump MAXMSG up to 2048
- Refactor serialize_state to use no more bytes that necessary, rather than
  using MAXMSG as before
- Server now sends reply to user registration to indicate success
- Remove unnecessary and barely-utilized Player->io_lock, since all writes are
  done with a single call, which POSIX guarantees to be thread-safe.
- Add additional safety checks to renderer so that the program doesn't throw
  nasty errors (segfault) when the renderer has not been properly initialized
- Fix bug in read_from_client that referenced the buffer rather than the cursor

bitoffdev committed:
Big Tetris Parties

TL;DR This patch allows more than the previous 2 players to compete, and
makes many minor improvements.

- Add magic number to all messages to help catch any network issues
- Add message_type field to message wire header, and remove the message_type
  from the body of many messages
- Define new method, message_type_to_str, to print a human-readable message
  type for logging
- Create new helper method message_blob
- Declare a macro name "PLAYER_NAME_MAX_CHARS", and use it consistently to
  validate the player names.
- Refactor StringArray to take a max length and define the StringArray
  serialized header bytes using a struct
- Bump MAXMSG up to 2048
- Refactor serialize_state to use no more bytes that necessary, rather than
  using MAXMSG as before
- Server now sends reply to user registration to indicate success
- Remove unnecessary and barely-utilized Player->io_lock, since all writes are
  done with a single call, which POSIX guarantees to be thread-safe.
- Add additional safety checks to renderer so that the program doesn't throw
  nasty errors (segfault) when the renderer has not been properly initialized
- Fix bug in read_from_client that referenced the buffer rather than the cursor

Nitepone commented on src/client_conn.c:
```diff
@@ -54,55 +56,50 @@ int get_socket() {
-	message[0] = MSG_TYPE_LIST;
-	NetRequest *request = ttetris_net_request(net_client, message, 1);
+	NetRequest *request =
+	    ttetris_net_request(net_client, "", 0, MSG_TYPE_LIST);
```
Curious. 2nd arg. NULL vs an empty string? Both seem to be used, is there a reason?

bitoffdev replied:
Ahh; good catch. I'll update this to use NULL since an empty string is unnecessary bloat.

Nitepone commented on src/client_conn.c:
```diff
@@ -112,43 +109,47 @@ void tetris_connect(NetClient *net_client, char *host, int port) {
 	// get the player associated with the board
 	char *board_name = buffer;
-	int name_length = strlen(board_name);
+	unsigned long name_length = strnlen(board_name, PLAYER_NAME_MAX_CHARS);
```
NIT: Agree with unsigned usage. But I hope that name length doesn't exceed the size of an int. Is there a reason to use a long here?

bitoffdev replied:
That's fair; I used a long because `strnlen` returns a `size_t`, which on most systems equates to an unsigned long afaik.

Of course, in practice `strnlen` will return a number smaller than `PLAYER_NAME_MAX_CHARS`, which is currently equal to 15, so I guess we could even use an `unsigned char`.

I will use `unsigned int` for now, per your suggestion. :smile:

Nitepone replied:
That is interesting, didn't look at the output of strnlen!

Nitepone commented on src/event.h:
```diff
@@ -0,0 +1,28 @@
+/**
+ * simple library for describing events and waiting for events
```
This seems to currently be used primarily for game start and game stop. Is there plans to do more this this?

bitoffdev replied:
You're correct that it's currently just for game start/stop, but it's super general. We could refactor some of the request-response networking code to use `event.h`. I am hoping it will be useful in the future as well, but I don't have any other specific ideas.

Nitepone replied:
I see, sounds good to me! I bet we will find more uses for this.

Nitepone commented on src/event.h:
```diff
@@ -0,0 +1,28 @@
+#ifndef TTETRIS_EVENT_H
+#define TTETRIS_EVENT_H
+
+typedef struct ttetris_event TetrisEvent;
```
NIT: Having the struct and typedef separate is not something we have done much in our code. We have been defining structs like this entirely in the header files with a typedef. I believe that doing it this way will prevent usage of `struct ttetris_event` in files that import this header? Are there any other differences?

bitoffdev replied:
Preventing the usage of `struct ttetris_event` was my intent. My thought was to keep the definition opaque and force users to interact with `TetrisEvent`s through the interface defined by the methods in `header.h`. I believe this will make it easier to change the underlying functionality in the future if we want.

I'm not aware of any other differences.

Nitepone replied:
I fully agree with this rationale.

Nitepone commented on src/client.c:
```diff
@@ -97,12 +144,53 @@ int run_online(char *host, int port, char *username, char *opponent) {
+		case 1:
+			run_online(host, port);
+			break;
+		case 2:
```
Add a TODO here, And we should open up an issue for this.

bitoffdev replied:
Good point. Implementing this can go under #21.

Nitepone commented on src/client.c:
```diff
@@ -59,34 +65,75 @@ int run_list_online_players(char *host, int port) {
  * run an online game of tetris
  */
-int run_online(char *host, int port, char *username, char *opponent) {
+int run_online(char *host, int port) {
```
NIT: A TUI prompt for an address and port if `-a` is not specified should be added.

I'll create an issue for it if you agree.

bitoffdev replied:
+1 Go ahead and create the issue; that makes sense to me! :smiley: 

In the future, I also envision that we might have a default server like ttetris.example.org if we get around to packaging, etc.

Nitepone replied:
See #33 

I do agree with this. Additionally, it may be cool to have a public server. However, we will need to take into account security risks before deploying such a thing.

Nitepone commented on src/client_conn.c:
```diff
@@ -54,55 +56,50 @@ int get_socket() {
 	// deserialize names
 	Blob *body = malloc(sizeof(Blob));
-	body->bytes = request->cursor + 1;
+	body->bytes = request->cursor;
```
Curious, what changed to not require this `+ 1`?

bitoffdev replied:
Yeah; this is kind of confusing. It gets into the nitty-gritty of our message wire format.

The reason is because the first byte of the **message body** used to indicate the type of message (one of the `#declare`s in header.h such as `MSG_TYPE_REGISTER`). In this patch, I **moved the message type into MessageHeader**, which defines the first few bytes of every message.

Since the message type has been removed from the body, we no longer need to increment the cursor past it (The cursor is automatically pushed to the start of the body by logic in `read_from_client` and `read_from_server`).

In the future, I'd love to document our wire format a little so it's easy to understand when people find our repo. It's evolved a lot since the first version.

bitoffdev replied:
If you're curious, here is the [corresponding server change](https://github.com/Nitepone/terminally-tetris/pull/30/files/88a277429027f9bbede8ecf12556cdd087616809#diff-cbf91a87d3f07d2ad455ab70d64c16f55b7dc82be33346014bf25f91296d1e9eL40-L50) that removes the extra byte from the message body.

Nitepone replied:
I see! Makes sense to me!

Nitepone commented on src/generic.c:
```diff
@@ -62,36 +71,45 @@ char *string_array_get_item(StringArray *arr, int index) {
-
-	for (int i = 0; i < length; i++) {
-		int string_length = strlen(blob->bytes + message_index);
+	// read the header first
```
Some of this is getting a little complex. It might not be a bad idea to document how bits/bytes are used in these blobs?

bitoffdev replied:
Yeah; that's fair. I'll create an issue for documenting the wire format.

The `StringArray` serializer might be the most complex that exists in our code right now, tbh. My hope is that introducing the [struct ttetris_string_array_wire_header](https://github.com/Nitepone/terminally-tetris/blob/88a277429027f9bbede8ecf12556cdd087616809/src/generic.c#L27) in this patch at least clears up the first few bytes. I've started using structs more to increase readability around our wire format when practical.

bitoffdev replied:
Created: #31

Nitepone reviewed (commented):
Some questions before approving this for merge. But, I don't see anything urgent to change.

bitoffdev committed:
Address comments on PR

- Add TODO for controls menu
- Use unsigned int for string length in read_game_view_data
- Replace unnecessary empty string in client_conn.c with NULL

bitoffdev commented:
Thanks for taking a look at my giant diff @Nitepone! :raised_hands: :heart_eyes: 

I think this latest commit should address your comments.

Nitepone reviewed (approved):
LGTM
